BH-1095: Allow Developers to change option names

anthonyburchell opened the pull request:
## Description

This PR adds multiple choice fields back into ACM and adds better validation and a data structure that allows for content model creators to change choice names. The key value way we're storing them now should ensure that the slug is maintained and the name is dynamic.

anthonyburchell committed:
checkpoint: adds validation for option slugs in a ghost field

anthonyburchell committed:
use individual input fields for option slug

nickcernis committed:
fix: give new choice API IDs readOnly=false

So that existing fields have read-only API IDs,
but newly added fields have editable API IDs.

nickcernis committed:
fix: improve function name

anthonyburchell committed:
fix:data structure of saved fields and returned rest data

anthonyburchell committed:
add: validation for duplicate slugs and blank slug inputs

anthonyburchell committed:
adds slug identifier validation

anthonyburchell committed:
Merge branch 'main' into bh-1095-slugify-options

anthonyburchell committed:
Merge branch 'bh-1095-slugify-options'

anthonyburchell committed:
adds: working e2e tests

anthonyburchell committed:
adds more tests for choice removal

anthonyburchell committed:
revises tests and fixes doc block for slug validation

matthewguywright commented on includes/publisher/js/src/components/Field.jsx:
```diff
@@ -196,19 +196,23 @@ function fieldMarkup(field, modelSlug, errors, validate) {
-											)
+											field.value.some(function (slug) {
+												return (
+													Object.keys(slug) ==
```
=== vs ==

matthewguywright commented on includes/settings/js/src/components/fields/MultipleChoiceFields.jsx:
```diff
@@ -40,17 +54,17 @@ function MultipleChoiceFields({
 									<div
 										key={item.id}
-										className="field multiple-option-container-single"
+										className={`choices[${index}].container field multiple-option-container-single`}
```
That's an interesting class name. Do we want '.'s in the class name? Like would this be something like:
choices[1].container field ...

Or is this a WP standard thing?

anthonyburchell replied:
it surprisingly works, but yeah I think we could switch these to `-` between each word. `choices-1-container`

nickcernis replied:
If we're using that class only as a selector for the “remove choice” e2e test, another option is to:

1. Remove the className here:

```suggestion
										className={`field multiple-option-container-single`}
```
2. Change the test selector:

```diff
diff --git a/tests/acceptance/CreateContentModelMultipleChoiceFieldCest.php b/tests/acceptance/CreateContentModelMultipleChoiceFieldCest.php
index 3fc3613..26fe3f3 100644
--- a/tests/acceptance/CreateContentModelMultipleChoiceFieldCest.php
+++ b/tests/acceptance/CreateContentModelMultipleChoiceFieldCest.php
@@ -28,7 +28,7 @@ class CreateContentModelMultipleChoiceFieldCest
         $I->wait(1);
         $I->seeInField('choices[0].name', 'dog');
         $I->see('Editing “Favorite Animal” Field', '.field-list');
-        $I->click('.remove-option', '.choices[0].container');
+        $I->click('Remove choice');
         $I->wait(1);
         $I->seeInField('choices[0].name', 'cat');
     }
```

The test runner will click the first button with the “Remove choice” text, which is the same button that was previously identified by `choices[0].container`.

matthewguywright commented on includes/settings/js/src/components/fields/Form.jsx:
```diff
@@ -210,6 +210,20 @@ function Form({ id, position, type, editing, storedData }) {
 						});
 					});
 				}
+				if (err.code === "wpe_option_slug_undefined") {
```
It would be nice to see these errors broken out into a separate object or an enum and use a switch statement or external function instead. There are quite a few of these now.

mindctrl replied:
Agreed, but I think we can tackle that in the future.

nickcernis commented on includes/settings/js/src/components/fields/MultipleChoiceFields.jsx:
```diff
@@ -7,6 +7,7 @@ import supportedFields from "./supportedFields";
 import AddIcon from "../../../../../components/icons/AddIcon";
 import TrashIcon from "../../../../../components/icons/TrashIcon";
 import Icon from "../../../../../components/icons";
+import { toValidApiId } from "./toValidApiId";
```
Just a heads up that Andrew's PR at https://github.com/wpengine/atlas-content-modeler/pull/214 changes the location of `toValidApiId`. (Just something to look out for in case that PR is merged first.)

nickcernis commented:
This is working well for me!

Apart from the above comments, I noticed that the `multipleChoice` field does not appear in WPGraphQL responses (I think because no type in `map_html_field_type_to_graphql_field_type` or resolver in `register_content_fields_with_graphql` are configured).

That might need to be a separate PR before we can ship this. (Unless we want to ship without WPGraphQL support and add it later?) REST support looks ok to me.

apmatthews commented on includes/publisher/lib/field-functions.php:
```diff
@@ -102,7 +102,7 @@ function sanitize_field( string $type, $value ) {
 				$options_object = [];
 				foreach ( $value as $option ) {
-					$options_object[] = key( $option );
+					$options_object[] = $option;
```
I may be totally missing something, but does this block actually change the value? The way it's written, I'm wondering if we need the `is_array()` conditional at all.

If we do want to transform this value somehow when it's an array, should we keep the `key()` function there as it was? Otherwise, we'll be saving option names. For instance, this is what would be saved for a multi-select field:

```
a:2:{i:0;a:1:{s:3:"two";s:3:"Two";}i:1;a:1:{s:5:"three";s:5:"Three";}}
```

(Unrelated to this PR, but it looks like we don't even use the value that gets returned from the `sanitize_field()` function 😬. I'll make a note to follow up on that!)

nickcernis replied:
I agree that we could remove or adapt it. It was left when Repeater fields were removed:

- https://github.com/wpengine/atlas-content-modeler/pull/124#pullrequestreview-671211033
- https://github.com/wpengine/atlas-content-modeler/pull/124#issuecomment-850533862

mindctrl commented:
When adding this field to a model, the cursor focus goes to the name input for Choice 1, skipping the Field name input which appears first.

![image](https://user-images.githubusercontent.com/1254870/128188775-4ed7cd23-41ed-4cf7-9917-6e53fc38b538.png)

nickcernis commented on includes/publisher/js/src/components/Field.jsx:
```diff
@@ -196,19 +196,23 @@ function fieldMarkup(field, modelSlug, errors, validate) {
+										name={`atlas-content-modeler[${modelSlug}][${field.slug}][${index}][${item.slug}]`}
+										id={`atlas-content-modeler[${modelSlug}][${field.slug}][${index}][${item.slug}]`}
 										placeholder="Option Name"
+										value={item.name}
```
Agreed, good catch.

anthonyburchell committed:
merges main and splits option slug logic to appropriate places

anthonyburchell committed:
Update includes/settings/js/src/components/fields/MultipleChoiceFields.jsx

Co-authored-by: Andrew Matthews <andrew.matthews@wpengine.com>

anthonyburchell committed:
adds basic graphql responses

anthonyburchell committed:
Merge remote-tracking branch 'origin' into bh-1095-slugify-options-rebase

anthonyburchell committed:
merges main

Commit:
We can return a list

anthonyburchell committed:
return choice values as array of strings

anthonyburchell committed:
adds logic for displaying lists and single items in multi select

anthonyburchell committed:
Update includes/settings/js/src/components/fields/MultipleChoiceFields.jsx

Co-authored-by: Matthew Wright <1815200+matthewguywright@users.noreply.github.com>

anthonyburchell committed:
php linting fixes

anthonyburchell committed:
Merge branch 'bh-1095-slugify-options-rebase' of github.com:wpengine/atlas-content-modeler into bh-1095-slugify-options-rebase

anthonyburchell committed:
remove register from fields to prevent focus

anthonyburchell committed:
use item.slug not name for value of an input

nickcernis reviewed (commented):
@anthonyburchell I'm currently unable to create a multiple choice field. I see “Must create a choice first” when clicking Create, even after creating choices:

<img width="870" alt="Screenshot 2021-10-25 at 10 57 49" src="https://user-images.githubusercontent.com/647669/138666353-83089a04-e35c-4912-9d37-30356a2adf4a.png">

Guessing that may be responsible to some of the e2e test failures on this branch?

I left another comment above about one of Andrew's suggestions, and dismissed some stale comments.

anthonyburchell committed:
removes unused supportedFields function and changes display of MC to beta

anthonyburchell committed:
add shouldFocus property on initialized field

anthonyburchell committed:
prettier

anthonyburchell committed:
adds container to better target remove option button and fixes tests

anthonyburchell committed:
adds container to better target remove option button and fixes tests

anthonyburchell committed:
adds time to wait before field renders in tests

anthonyburchell committed:
wait longer

anthonyburchell committed:
wait even longer

anthonyburchell committed:
fix shouldfocus property

anthonyburchell committed:
fix shouldfocus property

anthonyburchell committed:
no need to wait that long

anthonyburchell committed:
remove useless array check

ChrisWiegman reviewed (commented):
We've chatted about this elsewhere but I feel it is important to bring it here before I approve this.

What about hiding the slug field for each choice? We're not using them in GraphQL and the requirement to create one manually is strange and confusing in the settings app. I feel we would be better serving users by autogenerating that in the background. How we do that, here or another ticket, is something I'm more than willing to discuss.

ChrisWiegman reviewed (dismissed):
Beyond the comments, I would like to see some tests for a GraphQL response (where appropriate).

anthonyburchell committed:
simplify error handling and add translatable strings

anthonyburchell committed:
breaks error clearing out of onChange function inline

anthonyburchell committed:
fix publisher side rendering and saving of multiple choice

anthonyburchell committed:
align spacing of conditions in field functions

anthonyburchell committed:
fix: single radio needs to compare slugs to decide default selected

anthonyburchell committed:
check for array before pushing to one

mindctrl commented:
@anthonyburchell can you bring this up to date with `main`? Looks like quite a few things are missing, like recent Parcel fixes.

anthonyburchell committed:
Merge remote-tracking branch 'origin/main' into bh-1095-slugify-options-rebase

mindctrl commented:
Given a model like this screenshot:
![image](https://user-images.githubusercontent.com/1254870/141845890-b599903a-e18a-4600-920e-1d78bef7ea92.png)

---

And adding an entry like this screenshot:
![image](https://user-images.githubusercontent.com/1254870/141845937-5507715d-2619-4eac-a194-6be1d0b61d03.png)

---

Results in the meta value being stored like this:
`a:1:{i:0;a:1:{s:10:"greenBeans";s:10:"greenBeans";}}`

Does this seem right? Do we need to store the value twice for each selection? If I add another selection, the value looks like this:
`a:2:{i:0;a:1:{s:10:"greenBeans";s:10:"greenBeans";}i:1;a:1:{s:4:"rice";s:4:"rice";}}`

matthewguywright commented:
General Notes:

1. Remove choice is awkward, especially in smaller screen resolutions for tablets and smaller. Is the text necessary? Can it use just the red trash can or a red "x"?
2. Currently it doesn't auto generate the identifier
3. Character count for "Name" field is misaligned with the input
4. Label not aligned vertically in publisher with the radio and checkbox input
5. Legend label looks indented by 2px from the rest of the form

anthonyburchell committed:
fix: saving slugs was creating a meta entry with duplicate slugs

anthonyburchell committed:
i8n multiple choice errors and add text domain

nickcernis commented on includes/rest-api/routes/content-model-field.php:
```diff
@@ -126,14 +125,69 @@ function dispatch_update_content_model_field( WP_REST_Request $request ) {
+		if ( $problem_option_slug_index ) {
+			$problem_duplicate_slug = new WP_Error(
+				'wpe_option_slug_duplicate',
+				'Another choice in this field has the same API identifier.',
```
Confirmed fixed in 44440f7c30a81f1d4045b153a22a8075f2e4bb2a.

nickcernis commented on includes/rest-api/routes/content-model-field.php:
```diff
@@ -126,14 +125,69 @@ function dispatch_update_content_model_field( WP_REST_Request $request ) {
+		if ( $problem_name_index ) {
+			$problem_duplicate_name = new WP_Error(
+				'wpe_duplicate_content_model_multi_option_id',
+				'Another choice in this field has the same name.',
```
Confirmed fixed in 44440f7c30a81f1d4045b153a22a8075f2e4bb2a.

nickcernis commented on includes/rest-api/routes/content-model-field.php:
```diff
@@ -126,14 +125,69 @@ function dispatch_update_content_model_field( WP_REST_Request $request ) {
+		if ( ! empty( $problem_index ) ) {
+			$problem_error_slug_blank = new WP_Error(
+				'wpe_option_slug_undefined',
+				'Multiple Choice Field update failed, please set a slug for your choice before saving.',
```
Confirmed fixed in 44440f7c30a81f1d4045b153a22a8075f2e4bb2a.

mindctrl commented on includes/rest-api/routes/content-model-field.php:
```diff
@@ -104,43 +105,96 @@ function dispatch_update_content_model_field( WP_REST_Request $request ) {
 		return new WP_Error(
 			'wpe_invalid_multi_options',
-			'Multiple Choice update failed. Options need to be created before updating a Multiple Choice field.',
+			__( 'Multiple Choice update failed. Options need to be created before updating a Multiple Choice field.', 'atlas-content-modeler' ),
```
```suggestion
			__( 'Multiple Choice update failed. Choices need to be created before updating a Multiple Choice field.', 'atlas-content-modeler' ),
```

Should we say "Choices" instead of "Options" here, since the UI calls them choices?

anthonyburchell committed:
uses single selects as an array to allow a single to become a multi

anthonyburchell committed:
Fix: graphql responses for multiple choice fields after meta value change

anthonyburchell committed:
add: graphql tests

anthonyburchell committed:
add linting

nickcernis committed:
test: use array instead of serialized string

nickcernis commented on includes/settings/js/src/components/fields/MultipleChoiceFields.jsx:
```diff
@@ -167,8 +270,10 @@ function MultipleChoiceFields({
-														option names.
+														{__(
+															"Cannot have duplicate choice names.",
+															"atlas-content-modeler"
```
I agree with this, but perhaps to keep the PR on track we could handle updating error message handling in a separate PR? There are some other places we should change it, as @anthonyburchell noted in standup.

nickcernis committed:
Merge branch 'main' into bh-1095-slugify-options-rebase

nickcernis committed:
chore: use “Choices” for consistency

Co-authored-by: John Parris <john.parris@wpengine.com>

nickcernis committed:
chore: remove unused variable

Co-authored-by: John Parris <john.parris@wpengine.com>

nickcernis committed:
[NOJIRA] Add data about duplicate multiple choice option slugs to the error directly (#361)

* refactor: add data about the duplicates

Removing the need to call 'add'. We just add info
about the duplicates to the error directly.

* refactor: pull data about the duplicates

Using the new `duplicates` property that we
set in the error message data.

anthonyburchell committed:
Add data about problem choice indexes directly from WP_Error

anthonyburchell committed:
remove text next to remove choice button and add aria-label

nickcernis commented on includes/rest-api/routes/content-model-field.php:
```diff
@@ -104,43 +105,96 @@ function dispatch_update_content_model_field( WP_REST_Request $request ) {
+				__( 'Another choice in this field has the same name.', 'atlas-content-modeler' ),
+				array( 'status' => 400 )
+			);
+			$problem_duplicate_name->add( 'problem_name_index', $problem_name_index );
```
Addressed in 866bbfb35be5dcc76f7d21d029279fc8d43f9935 and 2543ab9dc6c5a98e8aa98dcba9c86c513a3a4518.

nickcernis committed:
style: adjust legend spacing

nickcernis committed:
style: center publisher radio button labels

nickcernis reviewed (dismissed):
Looks good! 🥳 

Noting that we chose not to auto-generate the choice slugs from their names for now. (This requires a refactor of the `useInputGenerator` hook to support dynamically sized arrays of inputs, which seems out of scope for now.)

anthonyburchell committed:
fix: return empty array when no choices are saved for an entry
